kep/108-Memory-bandwidth-and-LLC-aware-filtering-and-scoring/README.md

```diff
@@ -0,0 +1,145 @@
+# Memory bandwidth and LLC aware filtering and scoring
+
+## Table of Contents
+
+<!-- toc -->
+- [RDT aware filtering and scoring](#RDT-aware-filtering-and-scoring)
+  - [Table of Contents](#table-of-contents)
+  - [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Terms](#terms)
+  - [Proposal](#proposal)
+  - [Design Details](#design-details)
+    - [Extension points](#extension-points)
+      - [Filter](#filter)
+      - [PreScore](#prescore)
+      - [Score](#score)
+  - [Known Limitations](#known-limitations)
+  - [Alternatives considered](#alternatives-considered)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Testing Plan](#testing-plan)
+  - [Implementation History](#implementation-history)
+  - [References](#references)
+<!-- /toc -->
+
+## Motivation
+
+Memory bandwidth and LLC (last level cache) are very important resources in a system. Kuberenetes scheduler hasn't taken them into consideration yet. But memory bandwidth and LLC contention may impact system performance very much. Furtherly, it causes bad SLA. Intel Resource Director Technology (RDT) is one example of the technology that can be leveraged to mitigate the impact of memory bandwidth and LLC contention. It is very helpful to ensure system performance stability and improve SLA. 
+
+## Goals
+1. Use scheduler plugin, which is the most Kubernetes native way, to implement memory bandwidth and LLC aware filtering and scoring.
+2. Leverage memory bandwidth and LLC relative metrics for filtering and scoring.
+3. Provide configurable weigths for prioritizing the metrics used in the scoring calculations.
+
+## Non-Goals
+
+## Use Cases
+
+## Terms
+LLC: last level cache.
+RDT: Intel Resource Director Technology.
+SLA: Service Level Agreement.
+MPKI: Misses Per Kilo Instructions.
+
+## Proposal
+We introduce some metrics such as memory bandwidth utilization, LLC occupation etc. At schedule stage, we can leverage these metrics to assist filting and scoring. It is hard to determine how important a resource is and how much resource is required. We did off-line analysis to different scenario and get a series of coefficient of correlation and generic usage value. The coefficient is called affinity and the generic usage value is called profile. The affinities and profiles are provided by a yaml configuration file. Detail about how the affinity and profile are used is described in the following design details section. The average value of a period of time is more meaningful than immediate value for some metrics, for instance, LLC occupancy and CPU usage. Metrics data aren't obtained from agents (such as cAdvisor) directly, instead they are stored in Prometheus first and then are retrieved on demand.
+
+## Design Details
+The filtering and scoring plugin leverages the metrics retrieved from Prometheus to determine the node candidates for running a pod. The metrics considered by the plugin at the moment are:
```

Comment:
What will happening if the Prometheus metrics are not available at scheduling time?

Reply:
Also, what is he source of the prometheus metrics(cadvisor, node_exporter)? Is a custom Prometheus exporter required to collect these metrics?

Reply:
If Prometheus metrics are not available it will let all node pass the filter and return the same score.
Currently it‘s WCA (https://github.com/intel/workload-collocation-agent). Some patches for cadvisor have been submitted,  it will be cadvisor once the patches are merged.

Reply:
The [real load aware scheduling KEP](https://github.com/kubernetes-sigs/scheduler-plugins/tree/master/kep/61-Trimaran-real-load-aware-scheduling) mentioned the use of a load watcher to cache metrics that the scheduler could then use.

I'm not saying the same approach needs to be used in the plugins proposed here, but just wanted to let you know that another set of plugins is also planning to leverage "real metrics".

Reply:
I think this proposal is different with https://github.com/kubernetes-sigs/scheduler-plugins/tree/master/kep/61-Trimaran-real-load-aware-scheduling 
This proposal will consider richer metrics in scheduling not only the CPU utilization.

I agree with you. We can consider reusing metrics as much as possible in different plugins. @seanmalloy

Reply:
To be clear I'm not saying that the metrics should be reused between these two plugins. The metrics used by these two plugins are completely different.

I'm saying we might want to consider having a common way for the scheduler code to access the metrics. If I understand correctly this proposal is suggesting directly fetching metrics from a prometheus server. The real load aware scheduling proposal has suggested a different method for the scheduler code to access metrics.

Reply:
I see. Thank you for your detailed explanation